src/Microsoft.Extensions.FileProviders.Physical/ConcurrentLruCache.cs

```diff
@@ -0,0 +1,247 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.FileProviders.Physical
+{
+    /// <summary>
+    /// Cache with a fixed size that evicts the least recently used members.
+    /// Thread-safe.
+    /// </summary>
+    internal class ConcurrentLruCache<K, V>
+    {
+        private readonly int _capacity;
+
+        private struct CacheValue
+        {
+            public V Value;
+            public LinkedListNode<K> Node;
+        }
+
+        private readonly Dictionary<K, CacheValue> _cache;
+        private readonly LinkedList<K> _nodeList;
+        // This is a naive course-grained lock, it can probably be optimized
+        private readonly object _lockObject = new object();
+
+        public ConcurrentLruCache(int capacity)
+        {
+            if (capacity <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity));
+            }
+            _capacity = capacity;
+            _cache = new Dictionary<K, CacheValue>(capacity);
+            _nodeList = new LinkedList<K>();
+        }
+
+        /// <summary>
+        /// Create cache from an array. The cache capacity will be the size
+        /// of the array. All elements of the array will be added to the 
+        /// cache. If any duplicate keys are found in the array a
+        /// <see cref="ArgumentException"/> will be thrown.
+        /// </summary>
+        public ConcurrentLruCache(KeyValuePair<K, V>[] array)
+            : this(array.Length)
+        {
+            foreach (var kvp in array)
+            {
+                this.UnsafeAdd(kvp.Key, kvp.Value, true);
+            }
+        }
+
+        /// <summary>
+        /// For testing. Very expensive.
+        /// </summary>
+        internal IEnumerable<KeyValuePair<K, V>> TestingEnumerable
+        {
+            get
+            {
+                lock (_lockObject)
+                {
+                    var copy = new KeyValuePair<K, V>[_cache.Count];
+                    int index = 0;
+                    foreach (K key in _nodeList)
+                    {
+                        copy[index++] = new KeyValuePair<K, V>(key,
+                            _cache[key].Value);
+                    }
+                    return copy;
+                }
+            }
+        }
+
+        public void Add(K key, V value)
+        {
+            lock (_lockObject)
+            {
+                UnsafeAdd(key, value, true);
+            }
+        }
+
+        private void MoveNodeToTop(LinkedListNode<K> node)
+        {
+            if (!object.ReferenceEquals(_nodeList.First, node))
+            {
+                _nodeList.Remove(node);
+                _nodeList.AddFirst(node);
+            }
+        }
+
+        /// <summary>
+        /// Expects non-empty cache. Does not lock.
+        /// </summary>
+        private void UnsafeEvictLastNode()
+        {
+            var lastNode = _nodeList.Last;
+            _nodeList.Remove(lastNode);
```

Comment:
'_nodeList.RemoveLast()'? Otherwise it will search in the whole list to find the last.